shell/platform/android/io/flutter/plugin/editing/SpellCheckPlugin.java

```diff
@@ -0,0 +1,134 @@
+import io.flutter.plugin.localization.LocalizationPlugin;
+import java.util.ArrayList;
+import java.util.Locale;
+
+/**
+ * {@link SpellCheckPlugin} is the implementation of all functionality needed for spell check for
+ * text input.
+ *
+ * <p>The plugin handles requests for spell check sent by the {@link
+ * io.flutter.embedding.engine.systemchannels.SpellCheckChannel} via sending requests to the Android
+ * spell checker. It also receive the spell check results from the service and send them back to the
+ * framework through the {@link io.flutter.embedding.engine.systemchannels.SpellCheckChannel}.
+ */
+public class SpellCheckPlugin
+    implements SpellCheckChannel.SpellCheckMethodHandler,
+        SpellCheckerSession.SpellCheckerSessionListener {
+
+  private final SpellCheckChannel mSpellCheckChannel;
+  private final TextServicesManager mTextServicesManager;
+  private SpellCheckerSession mSpellCheckerSession;
+
+  // The maximum number of suggestions that the Android spell check service is allowed to provide
+  // per word.
+  // Same number that is used by default for Android's TextViews.
+  private static final int MAX_SPELL_CHECK_SUGGESTIONS = 5;
+
+  public SpellCheckPlugin(
+      @NonNull TextServicesManager textServicesManager,
+      @NonNull SpellCheckChannel spellCheckChannel) {
+    mTextServicesManager = textServicesManager;
+    mSpellCheckChannel = spellCheckChannel;
+
+    mSpellCheckChannel.setSpellCheckMethodHandler(this);
+  }
+
+  /**
+   * Unregisters this {@code SpellCheckPlugin} as the {@code
+   * SpellCheckChannel.SpellCheckMethodHandler}, for the {@link
+   * io.flutter.embedding.engine.systemchannels.SpellCheckChannel}, and closes the most recently
+   * opened {@code SpellCheckerSession}.
+   *
+   * <p>Do not invoke any methods on a {@code SpellCheckPlugin} after invoking this method.
+   */
+  public void destroy() {
+    mSpellCheckChannel.setSpellCheckMethodHandler(null);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+  }
+
+  @Override
+  public void initiateSpellCheck(@NonNull String locale, @NonNull String text) {
+    performSpellCheck(locale, text);
+  }
+
+  /** Calls on the Android spell check API to spell check specified text. */
+  public void performSpellCheck(@NonNull String locale, @NonNull String text) {
+    String[] localeCodes = locale.split("-");
+    Locale localeFromString = LocalizationPlugin.localeFromString(locale);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+    mSpellCheckerSession =
+        mTextServicesManager.newSpellCheckerSession(
+            null,
+            localeFromString,
+            this,
+            /** referToSpellCheckerLanguageSettings= */
+            true);
+
+    TextInfo[] textInfos = new TextInfo[] {new TextInfo(text)};
+    mSpellCheckerSession.getSentenceSuggestions(textInfos, MAX_SPELL_CHECK_SUGGESTIONS);
```

Comment:
Do you know what happens if you call getSentenceSuggestions multiple times before it has called back to onGetSentenceSuggestions?

Reply:
From my observation, it seems to consume calls and only output the results of the last request when a bunch of requests are sent back to back. This seems to happen when you're typing pretty fast and as the text getting spell checked gets longer, but I believe this isn't cause for concern, as long as we get the results of the most recent call and this is enough to not affect the visual of misspelled words. I have not seen that yet, but I think this point gives further motivation for optimizing this down the line by not spell checking the entirety of the text each time.